src/rwr.py

```diff
@@ -0,0 +1,155 @@
+        @param network:  input network file (required)
+        @param output_file: path to the output pathway file (required)
+        @param df: damping factor for restarting (default 0.85) (optional)
+        @param f: selection function (default 'min') (optional)
+        @param threshold: threshold for constructing the final pathway (default 0.0001) (optional)
+        @param singularity: if True, run using the Singularity container instead of the Docker container
+        """
+
+        if not edges or not sources or not targets or not output_file:
+            raise ValueError('Required RandomWalk arguments are missing')
+
+        work_dir = '/spras'
+
+        # Each volume is a tuple (src, dest) - data generated by Docker
+        volumes = list()
+
+        bind_path, edges_file = prepare_volume(edges, work_dir)
+        volumes.append(bind_path)
+
+        bind_path, sources_file = prepare_volume(sources, work_dir)
+        volumes.append(bind_path)
+
+        bind_path, targets_file = prepare_volume(targets, work_dir)
+        volumes.append(bind_path)
+
+
+        out_dir = Path(output_file).parent
+        # RandomWalk requires that the output directory exist
+        out_dir.mkdir(parents=True, exist_ok=True)
+        bind_path, mapped_out_dir = prepare_volume(str(out_dir), work_dir)
+        volumes.append(bind_path)
+        mapped_out_prefix= mapped_out_dir + '/out'  # Use posix path inside the container
+
+
+        command = ['python',
+                   '/RandomWalk/random_walk.py',
+                   '--edges_file', edges_file,
+                   '--sources_file', sources_file,
+                   '--targets_file', targets_file,
+                   '--damping_factor', str(df),
+                   '--selection_function', f,
+                   '--threshold', str(threshold),
+                   '--output_file', mapped_out_prefix]
+
+        print('Running RandomWalk with arguments: {}'.format(' '.join(command)), flush=True)
+
+
+        container_framework = 'singularity' if singularity else 'docker'
+        out = run_container(container_framework,
+                            'erikliu24/rwwr',
+                            command,
+                            volumes,
+                            work_dir)
+        print(out)
+
+        output = Path(out_dir, 'out')
+        output.rename(output_file)
+
+        # From edge_output_file, construct a pathway file in the universal format
+        # 1. Stop when the source and targets are connected.
+
+    @staticmethod
+    def parse_output(raw_pathway_file, standardized_pathway_file):
+        """
+        Convert a predicted pathway into the universal format
+        @param raw_pathway_file: pathway file produced by an algorithm's run function
+        @param standardized_pathway_file: the same pathway written in the universal format
+        """
+        print('Parsing random-walk-with-restart output')
```

Comment:
is this needed?

Reply:
I had initially considered printing this line to the console to provide users with additional information about the progress of the program. However, I can remove this line if it would be more in line with the unified coding style of SPRAS.

Reply:
We can remove this parsing output message. We do some logging when it is helpful for debugging, like the Python commands that are going to be run inside the Docker containers. Otherwise for general information about workflow progress we rely on the Snakemake output.